pkg/pool-weighted/contracts/BaseWeightedPool.sol

```diff
@@ -280,21 +290,28 @@ abstract contract BaseWeightedPool is BaseMinimalSwapInfoPool, WeightedMath {
         uint256[] memory balances,
         uint256[] memory normalizedWeights,
         bytes memory userData
-    ) private view returns (uint256, uint256[] memory) {
+    ) private returns (uint256, uint256[] memory) {
         (uint256 bptAmountOut, uint256 tokenIndex) = userData.tokenInForExactBptOut();
         // Note that there is no maximum amountIn parameter: this is handled by `IVault.joinPool`.
 
         _require(tokenIndex < _getTotalTokens(), Errors.OUT_OF_BOUNDS);
 
-        uint256[] memory amountsIn = new uint256[](_getTotalTokens());
-        amountsIn[tokenIndex] = WeightedMath._calcTokenInGivenExactBptOut(
+        (uint256 amountIn, uint256 swapFee) = WeightedMath._calcTokenInGivenExactBptOut(
             balances[tokenIndex],
             normalizedWeights[tokenIndex],
             bptAmountOut,
             totalSupply(),
             getSwapFeePercentage()
         );
 
+        // Note that swapFee is already upscaled
+        _processSwapFeeAmount(tokenIndex, swapFee);
+
+        // We join in a single token, so we initialize amountsIn with zeros
+        uint256[] memory amountsIn = new uint256[](_getTotalTokens());
+        // And then assign the result to the selected token
+        amountsIn[tokenIndex] = amountIn;
```

Comment:
Any reason for the difference here and in WeightedPool2Tokens? https://github.com/balancer-labs/balancer-v2-monorepo/blob/36c3847a0dd0cd9f3e98fa7130414153c4a58fe0/pkg/pool-weighted/contracts/WeightedPool2Tokens.sol#L574-L581

Reply:
Similarly for exits

Reply:
WeightedPool2Tokens ignores the swap fee, so I just kept the old code and assigned directly into the array. We could change those to make them as similar as possible.

Part of the reason is that Solidity lets you either assign to variables in tuple deconstruction or declare new ones, but not both. For example, the following is invalid:

```
function baz() returns (uint256, uint256);

uint256 foo;
(foo, uint256 bar) = baz();
```

Reply:
Ah, I missed the difference in handling of swap fees. In that case then I agree we should leave this as be as we won't really gain much in return for the reduced readability.